tap_exchangeratesapi.py

```diff
@@ -48,23 +49,31 @@ def request(url, params):
     
 def do_sync(base, start_date):
     logger.info('Replicating exchange rate data from exchangeratesapi.io starting from {}'.format(start_date))
-    singer.write_schema('exchange_rate', schema, 'date')
 
     state = {'start_date': start_date}
     next_date = start_date
+    prev_schema = {}
     
     try:
-        while True:
+        while datetime.strptime(next_date, DATE_FORMAT) <= datetime.utcnow():
             response = request(base_url + next_date, {'base': base})
             payload = response.json()
 
-            if datetime.strptime(next_date, DATE_FORMAT) > datetime.utcnow():
-                break
-            else:
-                if payload['date'] == next_date:
-                    singer.write_records('exchange_rate', [parse_response(payload)])
-                state = {'start_date': next_date}
-                next_date = (datetime.strptime(next_date, DATE_FORMAT) + timedelta(days=1)).strftime(DATE_FORMAT)
+            # Update schema if new currency/currencies exist
+            for rate in payload['rates']:
+                if rate not in schema['properties']:
+                    schema['properties'][rate] = {'type': ['null', 'number']}
+
+            # Only write schema if it has changed
+            if schema != prev_schema:
```

Comment:
@CStejmar Can schemas vary between days? If the API always returns the same currencies, or realistically within the same tap run, I would just emit the schema once on the first returned response.

Reply:
@awm33 Yes, it can vary between days. For example for EEK as I described above. I also noticed that ISK, for example, was added 2018-02-01. So if you would do a tap run from 2018-01-31 to 2018-02-01 you would not get the ISK into the database if you just emit the schema once. But yes, it does not change very often. What do you think, should I keep it as it is or change? I think it is more robust if we keep it as it is, but change `additionalProperties` to `False` as you suggested.

Reply:
@CStejmar Ah, ok, then I think updating the schema each call makes sense.

Reply:
Ok, I will go ahead and change the `additionalProperties` 👍